### ISSUE TITLE
Preconditions and Post-operation validation steps

### ISSUE THREAD

**Issue by wookayin** [Fizzadar/pyinfra#367]
What is the best way to check preconditions? Because of the [*fundamental limitation of the execution model*](https://docs.pyinfra.com/en/latest/examples/dynamic_execution_deploy.html), we must use `facts` which are *statically* gathered before the execution.

But often the case we may want to check preconditions for operations to be performed. **If the preconditions aren't met, the deploy should abort** and throw an error with a informative error message. For example, checking existence of a file/directory, etc. But what is the best practice for doing this? None of the examples has a *real-world* use case of such a need, so I couldn't come up with an answer.

Throwing exception would not be a good idea, because exceptions are exceptions. Is there any API I should use for doing this?

Also, what is the way for checking some conditions **at runtime**, after executions are done? This is most needed for post-operation validation (e.g. if the file indeed exists). Running `server.shell` might be an option but this is very inconvenient to do check and log the details (i.e. why).

**Comment by wookayin** [comment]
Looks like one should [write a custom operation instead](https://docs.pyinfra.com/en/latest/api/operations.html), which is a minimal idempotent unit. This seems to give a full control of flow and checking status of the host dynamically.

**Comment by Fizzadar** [comment]
> But often the case we may want to check preconditions for operations to be performed. If the preconditions aren't met, the deploy should abort and throw an error with a informative error message. For example, checking existence of a file/directory, etc. But what is the best practice for doing this?

This is a really interesting idea! Perhaps this could be a global argument to execute some arbitrary shell before an operation, failing if it exits incorrectly. Eg:

```py
files.put(
    name='Upload this file',
    src='this-file',
    dest='/path/to/this-file',
    precondition=('stat /some/other/dir', '/some/other/dir not present, exiting!'),
)
```

I think that would work? I think a similar approach could also work for `postcondition`? Keen to hear your thoughts!

**Comment by Fizzadar** [comment]
I have now implemented a beta version of this @ https://github.com/Fizzadar/pyinfra/commit/fb6fb2eefbf7fb482a80a96f64455f460f25ef6d. Currently only accepts a single string command.

**Comment by wookayin** [comment]
Thanks for your support. I think support of pre/post-condition via shell commands will satisfy many use cases, but it'd be even better to have a pythonic API akin to `hosts.facts`. This is also because those operations can be arbitrary complex and we don't want to implement such logics using shell commands as in facts and operations.

However, that would be definitely an advanced feature so for now it should be fine. I will test this feature soon, hopefully before the release.

**Comment by Fizzadar** [comment]
100% agree - this is definitely a basic implementation and keen to expand upon this. Using facts is a really interesting idea, can definitely see re-using facts to verify operations as part of a `postcondition`!

**Comment by Fizzadar** [comment]
Now released in `v1.1.dev0`, hoping to get 1.1 out soon. Will close this issue for now and we can create subsequent issues for more complex implementations of pre/post conditions.
